src/template/js/sdk/Capabilities/index.mjs

```diff
@@ -18,43 +18,82 @@
 
 import Transport from '../Transport/index.mjs'
 import Events from '../Events/index.mjs'
-const validProviderMethods = {}
 
-export const registerProviderMethods = (module, methods) => {
-  validProviderMethods[module.toLowerCase()] = methods.concat()
-}
+const providerInterfaces = {}
+
+export const registerProviderInterface = (capability, module, methods) => {
+  if (providerInterfaces[capability]) {
+    throw `Capability ${capability} has multiple provider interfaces registered.`
+  }
 
-const getProviderArgs = function(...args) {
-  const provider = args.pop()
-  const module = args[0].toLowerCase() || '*'
-  return [module, provider]
+  methods.forEach(m => m.name = `${module}.${m.name}`)
+  providerInterfaces[capability] = methods.concat()
 }
 
-const provide = function(...args) {
-  const [module, provider] = getProviderArgs(...args)
-  const methods = Object.getOwnPropertyNames(Object.getPrototypeOf(provider)).filter(item => {
-    return typeof provider[item] === 'function' && item != 'constructor'
-  })
-  const pms = validProviderMethods[module]
-  for (let i = 0; i < methods.length; i++) {
-    const name = methods[i].charAt(0).toUpperCase() + methods[i].slice(1);
-    if (pms.indexOf(methods[i]) !== -1) {
-      Events.listen(module, 'request' + name, async function (req) {
-        const fn = provider[methods[i]]
-        const providerCallArgs = [req.request, resp => {
-          Transport.send(module, methods[i] + 'Response', {
-            correlationId: req.correlationId,
-            response: resp
+const provide = function(capability, provider) {
+  const methods = []
+  const iface = providerInterfaces[capability]
+
+  if (provider.constructor.name !== 'Object') {
+    methods.push(...Object.getOwnPropertyNames(Object.getPrototypeOf(provider)).filter(item => typeof provider[item] === 'function' && item !== 'constructor'))
+  }
+  else {
+    methods.push(...Object.getOwnPropertyNames(provider).filter(item => typeof provider[item] === 'function'))
+  }
+
+  if (iface) {
+
+    // make sure every interfaced method exists in the providers methods list
+    const valid = iface.every(method => methods.find(m => m === method.name.split('.').pop()))
+
+    if (valid) {
+      iface.forEach(imethod => {
+        const parts = imethod.name.split('.')
+        const method = parts.pop();
+        const module = parts.pop().toLowerCase();
+
+        const defined = !!methods.find(m => m === method) 
+
+        if (defined) {
+          Events.listen(module, `request${method.charAt(0).toUpperCase() + method.substr(1)}`, function (request) {
+
+            const providerCallArgs = []
+            
+            // only pass in parameters object if schema exists
+            if (imethod.parameters) {
+              providerCallArgs.push(request.parameters)
+            }
+
+            // only pass in the ready handshake if needed
+            if (imethod.handshake) {
```

Comment:
Did you swap this around where the ready callback is passed in and the promise resolution is the response?

Reply:
yup. see above thread.